narwhals/_pandas_like/group_by.py

```diff
@@ -40,231 +249,121 @@ def __init__(
         *,
         drop_null_keys: bool,
     ) -> None:
-        self._df = df
+        self._original_columns = tuple(df.columns)
         self._drop_null_keys = drop_null_keys
-        self._compliant_frame, self._keys, self._output_key_names = self._parse_keys(
-            df, keys=keys
+        ns = df.__narwhals_namespace__()
+        frame, self._keys, self._output_key_names = self._parse_keys(df, keys=keys)
+        self._exclude: tuple[str, ...] = (*self._keys, *self._output_key_names)
+        self._remap_non_str_columns = _remap_non_str(self)
+        self._compliant_frame = frame.with_columns(
+            *(ns.col(old).alias(new) for old, new in self._remap_non_str_columns.items())
         )
         # Drop index to avoid potential collisions:
         # https://github.com/narwhals-dev/narwhals/issues/1907.
-        if set(self.compliant.native.index.names).intersection(self.compliant.columns):
-            native_frame = self.compliant.native.reset_index(drop=True)
-        else:
-            native_frame = self.compliant.native
-
-        self._grouped = native_frame.groupby(
+        native = self.compliant.native
+        if set(native.index.names).intersection(self.compliant.columns):
+            native = native.reset_index(drop=True)
+        self._grouped: NativeGroupBy = native.groupby(
             list(self._keys),
             sort=False,
             as_index=True,
             dropna=drop_null_keys,
             observed=True,
         )
 
-    def agg(self, *exprs: PandasLikeExpr) -> PandasLikeDataFrame:  # noqa: C901, PLR0912, PLR0914, PLR0915
-        implementation = self.compliant._implementation
-        backend_version = self.compliant._backend_version
-        new_names: list[str] = self._keys.copy()
-
+    def agg(self, *exprs: PandasLikeExpr) -> PandasLikeDataFrame:
         all_aggs_are_simple = True
-        exclude = (*self._keys, *self._output_key_names)
+        agg_exprs: list[AggExpr] = []
         for expr in exprs:
-            _, aliases = evaluate_output_names_and_aliases(expr, self.compliant, exclude)
-            new_names.extend(aliases)
+            agg_exprs.append(AggExpr(expr).with_expand_names(self))
             if not self._is_simple(expr):
                 all_aggs_are_simple = False
 
-        # dict of {output_name: root_name} that we count n_unique on
-        # We need to do this separately from the rest so that we
-        # can pass the `dropna` kwargs.
-        nunique_aggs: dict[str, str] = {}
-        simple_aggs: dict[str, list[str]] = collections.defaultdict(list)
-        simple_aggs_functions: set[str] = set()
+        if all_aggs_are_simple:
+            result: pd.DataFrame
+            if named_aggs := self._named_aggs(agg_exprs):
```

Comment:
To contrast it with `_agg_complex`, I would rename `_named_aggs` to `_agg_simple`, then within it we call `e.named_aggs(self)`